src/compiler/crystal/tools/expand.cr

```diff
@@ -0,0 +1,185 @@
+module Crystal
+  class ExpandResult
+    JSON.mapping({
+      status:     {type: String},
+      message:    {type: String},
+      expansions: {type: Array(ExpandTrace), nilable: true},
+    })
+
+    def initialize(@status, @message)
+    end
+
+    def to_text(io)
+      io.puts message
+      expansions.try &.each_with_index do |expansion, i|
+        io.puts "expansion #{i + 1}:"
+        io << "   "
+        io.puts expansion.original.lines(chomp: false).join "   "
+        io.puts
+        expansion.expandeds.each_with_index do |expanded, j|
+          io << "~> "
+          io.puts expanded.lines(chomp: false).join "   "
+          io.puts
+        end
+      end
+    end
+  end
+
+  class ExpandTrace
+    JSON.mapping({
+      original:  {type: String},
+      expandeds: {type: Array(String)},
```

Comment:
`expanded` is also used to address a plural quantity, so you can just `s/expandeds/expanded` :)